File: impl/python/writ/objects.py
"""Signed objects: structural validation and signatures (sections 1.4 to 1.7, 2, 5, 6, 9.1).

verify_object() runs section 6.1 in its normative order and returns the
parsed object. signing_input() and identity() implement sections 1.4 and
1.5. Limits come from section 1.6, crit from section 1.7.
"""

from . import bounds as B
from .canon import canonicalize, is_int, parse, validate_value
from .errors import WritError
from .keys import is_b64u, is_did, sha256_b64u, verify as verify_sig

# Section 1.6
LIMITS = {"writ": 4096, "call": 65536, "tally": 262144, "revoke": 65536}
MAX_CHAIN = 8
MIN_NONCE_BYTES = 16
HASH_BYTES = 32
SIG_BYTES = 64

TYPES = ("writ", "call", "tally", "revoke")

# Every member this implementation understands, per type. A name in crit
# outside this set is reason unsupported_critical.
KNOWN_MEMBERS = {
    "writ": frozenset(["v", "typ", "iss", "hld", "bnd", "prv", "exp", "nnc", "crit", "sig"]),
    "call": frozenset(["v", "typ", "id", "chain", "from", "op", "args", "crit", "sig"]),
    "tally": frozenset(["v", "typ", "call", "writ", "op", "acc", "st", "err", "out",
                        "used", "rev", "sub", "wrt", "crit", "sig"]),
    "revoke": frozenset(["v", "typ", "writ", "iss", "chain", "crit", "sig"]),
}

TALLY_STATES = ("ok", "failed", "canceled", "pending")

SIGNER_MEMBER = {"writ": "iss", "call": "from", "revoke": "iss"}


# ------------------------------------------------------ identity and input

def signing_input(obj):
    """Section 1.4: <typ> "/" <v> 0x00 <canonical form without sig>."""
    body = {k: v for k, v in obj.items() if k != "sig"}
    prefix = f"{obj['typ']}/{obj['v']}".encode("utf-8")
    return prefix + b"\x00" + canonicalize(body)


def identity(obj):
    """Section 1.5: hash of the canonical form of the whole object, sig included."""
    return sha256_b64u(canonicalize(obj))


def hash_body(body):
    """Hash of the canonical form of a result body (tally.out)."""
    return sha256_b64u(canonicalize(body))


# ------------------------------------------------------------ 6.1 helpers

def _mal(msg):
    return WritError("malformed", msg)


def _is_hash(v):
    return isinstance(v, str) and is_b64u(v, nbytes=HASH_BYTES)


def _require_b64u(obj, name, allow_null=False, allow_star=False):
    """Section 1.1 rule 5 for one member, if it is present and a string."""
    if name not in obj:
        return
    v = obj[name]
    if v is None and allow_null:
        return
    if v == "*" and allow_star:
        return
    if isinstance(v, str) and not is_b64u(v):
        raise WritError("noncanonical", f"{name} is not unpadded base64url")


def _check_binary_members(obj, typ):
    """Rule 5 on the binary members that belong to this object type."""
    _require_b64u(obj, "sig")
    if typ == "writ":
        _require_b64u(obj, "prv", allow_null=True)
        _require_b64u(obj, "nnc")
    elif typ == "call":
        _require_b64u(obj, "id")
    elif typ == "tally":
        _require_b64u(obj, "call")
        _require_b64u(obj, "writ")
        _require_b64u(obj, "out", allow_null=True)
        err = obj.get("err")
        if isinstance(err, dict):
            _require_b64u(err, "ref")
    elif typ == "revoke":
        _require_b64u(obj, "writ", allow_star=True)


def _check_chain_length(obj):
    chain = obj.get("chain")
    if isinstance(chain, list) and len(chain) > MAX_CHAIN:
        raise WritError("too_large", f"chain has {len(chain)} writs, limit {MAX_CHAIN}")


def load(data, typ):
    """Steps 1 and 2 of section 6.1. Returns (object, canonical bytes).

    ``data`` is received bytes or text, or an already parsed value. For raw
    input the received length is checked against the limit before parsing,
    then the canonical length after. Chain length is checked here too, so
    that both limits precede any signature work.
    """
    if typ not in TYPES:
        raise ValueError(f"unknown object type {typ}")
    limit = LIMITS[typ]
    if isinstance(data, (bytes, str)):
        raw_len = len(data) if isinstance(data, bytes) else len(data.encode("utf-8", "surrogatepass"))
        if raw_len > limit:
            raise WritError("too_large", f"{typ} is {raw_len} bytes, limit {limit}")
        obj = parse(data)
    else:
        obj = data
        validate_value(obj)
    canon = canonicalize(obj)
    if len(canon) > limit:
        raise WritError("too_large", f"{typ} canonical form is {len(canon)} bytes, limit {limit}")
    if not isinstance(obj, dict):
        raise _mal(f"{typ} is not a JSON object")
    _check_binary_members(obj, typ)
    _check_chain_length(obj)
    return obj, canon


def _check_version_and_type(obj, typ):
    """Step 3."""
    v = obj.get("v")
    if not (is_int(v) and v == 1):
        raise WritError("unsupported_version", f"v is {v!r}, expected 1")
    if obj.get("typ") != typ:
        raise WritError("wrong_type", f"typ is {obj.get('typ')!r}, expected {typ!r}")


def _check_crit(obj, typ):
    """Step 4, section 1.7."""
    if "crit" not in obj:
        return
    crit = obj["crit"]
    if not isinstance(crit, list) or not all(isinstance(n, str) for n in crit):
        raise _mal("crit is not an array of strings")
    for name in crit:
        if name not in KNOWN_MEMBERS[typ]:
            raise WritError("unsupported_critical", f"crit names {name!r}, which this verifier does not understand")
        if name not in obj:
            raise _mal(f"crit names {name!r}, which is not present")


def _require(obj, name, pred, what):
    if name not in obj:
        raise _mal(f"required member {name} is missing")
    if not pred(obj[name]):
        raise _mal(f"member {name} is not {what}")
    return obj[name]


def _require_key(obj, name):
    v = _require(obj, name, lambda x: isinstance(x, str), "a string")
    if not is_did(v):
        raise WritError("bad_key", f"{name} is not an Ed25519 did:key: {v!r}")
    return v


def _require_sig(obj):
    v = _require(obj, "sig", lambda x: isinstance(x, str), "a string")
    if not is_b64u(v, nbytes=SIG_BYTES):
        raise _mal("sig is not 86 characters of base64url")
    return v


def _require_hash(obj, name, allow_null=False):
    if name not in obj:
        raise _mal(f"required member {name} is missing")
    v = obj[name]
    if v is None and allow_null:
        return v
    if not _is_hash(v):
        raise _mal(f"member {name} is not a 43 character hash")
    return v


def check_bnd(bnd):
    """Structural rules for a writ's bnd (sections 3 and 3.2)."""
    if not isinstance(bnd, dict):
        raise _mal("bnd is not an object")
    if "act" not in bnd:
        raise _mal("bnd lacks act")
    for name in sorted(bnd, key=lambda k: k.encode("utf-16-be", "surrogatepass")):
        B.validate(bnd[name], name)
    if bnd["act"]["t"] != "prefix":
        raise _mal("act is not a prefix bound")
    if "hld" in bnd:
        if bnd["hld"]["t"] != "set":
            raise _mal("hld bound is not a set")
        for e in bnd["hld"]["v"]:
            if not isinstance(e, str):
                raise _mal("hld set element is not a string")
            if not is_did(e):
                raise WritError("bad_key", f"hld set element is not an Ed25519 did:key: {e!r}")
    if "depth" in bnd and bnd["depth"]["t"] != "max":
        raise _mal("depth bound is not a max")


def _check_writ_members(obj):
    _require_key(obj, "iss")
    _require_key(obj, "hld")
    check_bnd(_require(obj, "bnd", lambda x: isinstance(x, dict), "an object"))
    _require_hash(obj, "prv", allow_null=True)
    _require(obj, "exp", is_int, "an integer")
    nnc = _require(obj, "nnc", lambda x: isinstance(x, str), "a string")
    if not is_b64u(nnc, min_bytes=MIN_NONCE_BYTES):
        raise _mal("nnc is shorter than 16 bytes")
    _require_sig(obj)


def _check_writ_array(obj, name):
    arr = _require(obj, name, lambda x: isinstance(x, list), "an array")
    for i, w in enumerate(arr):
        if not isinstance(w, dict):
            raise _mal(f"{name}[{i}] is not an object")
    return arr


def _check_call_members(obj):
    ident = _require(obj, "id", lambda x: isinstance(x, str), "a string")
    if not is_b64u(ident, min_bytes=MIN_NONCE_BYTES):
        raise _mal("id is shorter than 16 bytes")
    chain = _check_writ_array(obj, "chain")
    if not chain:
        raise _mal("chain is empty")
    _require_key(obj, "from")
    _require(obj, "op", lambda x: isinstance(x, str), "a string")
    _require(obj, "args", lambda x: isinstance(x, dict), "an object")
    _require_sig(obj)


def _check_tally_members(obj):
    _require_hash(obj, "call")
    _require_hash(obj, "writ")
    _require(obj, "op", lambda x: isinstance(x, str), "a string")
    _require(obj, "acc", is_int, "an integer")
    st = _require(obj, "st", lambda x: isinstance(x, str) and x in TALLY_STATES, "one of ok, failed, canceled, pending")
    if "err" not in obj:
        raise _mal("required member err is missing")
    err = obj["err"]
    if st == "ok":
        if err is not None:
            raise _mal("err must be null when st is ok")
    else:
        if not isinstance(err, dict) or not isinstance(err.get("code"), str):
            raise _mal("err must be an object with a string code when st is not ok")
        if "ref" in err and not _is_hash(err["ref"]):
            raise _mal("err.ref is not a hash")
    _require_hash(obj, "out", allow_null=True)
    used = _require(obj, "used", lambda x: isinstance(x, dict), "an object")
    for k, v in used.items():
        if not is_int(v) or v < 0:
            raise _mal(f"used[{k}] is not a non-negative integer")
    if "rev" not in obj:
        raise _mal("required member rev is missing")
    rev = obj["rev"]
    if rev is not None:
        if not isinstance(rev, dict) or not is_int(rev.get("until")):
            raise _mal("rev must be null or an object with an integer until")
    sub = _require(obj, "sub", lambda x: isinstance(x, list), "an array")
    for i, s in enumerate(sub):
        if not isinstance(s, dict):
            raise _mal(f"sub[{i}] is not an object")
    _check_writ_array(obj, "wrt")
    if st == "pending" and (used != {} or rev is not None or obj["out"] is not None):
        raise _mal("a pending tally must have used {}, rev null and out null")
    _require_sig(obj)


def _check_revoke_members(obj):
    if "writ" not in obj:
        raise _mal("required member writ is missing")
    w = obj["writ"]
    if w != "*" and not _is_hash(w):
        raise _mal("writ is neither a hash nor \"*\"")
    _require_key(obj, "iss")
    chain = _check_writ_array(obj, "chain")
    if w == "*" and chain:
        raise _mal("chain must be empty when writ is \"*\"")
    if w != "*" and not chain:
        raise _mal("chain must not be empty when writ is a hash")
    _require_sig(obj)


_MEMBER_CHECKS = {
    "writ": _check_writ_members,
    "call": _check_call_members,
    "tally": _check_tally_members,
    "revoke": _check_revoke_members,
}


def check_structure(data, typ):
    """Section 6.1 steps 1 to 5. Returns the parsed object."""
    obj, _ = load(data, typ)
    _check_version_and_type(obj, typ)
    _check_crit(obj, typ)
    _MEMBER_CHECKS[typ](obj)
    return obj


def check_signature(obj, signer):
    """Section 6.1 step 6 for an object that passed steps 1 to 5."""
    verify_sig(signer, signing_input(obj), obj["sig"])


def verify_object(data, typ, signer=None):
    """Section 6.1 in full. Returns the parsed object.

    ``signer`` is the did:key that must have signed. For a writ, call, or
    revoke it defaults to the object's own iss or from member. A tally's
    signer is the hld of the writ it names, which the caller must supply.
    """
    obj = check_structure(data, typ)
    if signer is None:
        member = SIGNER_MEMBER.get(typ)
        if member is None:
            raise ValueError("a tally's signer must be supplied (hld of the writ it names)")
        signer = obj[member]
    check_signature(obj, signer)
    return obj

File: impl/go/writ/objects.go
// Package writ implements the Writ protocol v0.1 objects: writ, call, tally,
// revoke, their parsing, chain attenuation, tally-tree verification, and
// issuance. It depends only on the jcs, keys, bound, and wire packages.
package writ

import (
	"encoding/json"
	"errors"
	"regexp"
	"strings"

	"writproto/bound"
	"writproto/jcs"
	"writproto/keys"
	"writproto/wire"
)

// Version is the only protocol version this package speaks.
const Version = 1

// Limits from spec section 1.6.
const (
	MaxChain       = 8
	MaxWritBytes   = 4096
	MaxCallBytes   = 65536
	MaxTallyBytes  = 262144
	MaxRevokeBytes = 65536
	MinRandom      = 22 // 16 bytes base64url
)

var b64Re = regexp.MustCompile(`^[A-Za-z0-9_-]+$`)

// Known member names per type, used only to decide whether a crit entry is understood.
var known = map[string]map[string]bool{
	"writ":   set("v", "typ", "iss", "hld", "bnd", "prv", "exp", "nnc", "crit", "sig"),
	"call":   set("v", "typ", "id", "chain", "from", "op", "args", "crit", "sig"),
	"tally":  set("v", "typ", "call", "writ", "op", "acc", "st", "err", "out", "used", "rev", "sub", "wrt", "crit", "sig"),
	"revoke": set("v", "typ", "writ", "iss", "chain", "crit", "sig"),
}

func set(names ...string) map[string]bool {
	m := map[string]bool{}
	for _, n := range names {
		m[n] = true
	}
	return m
}

// Writ is a parsed, structurally valid writ. Signature validity is separate.
type Writ struct {
	Raw wire.Object
	ID  string // identity hash, spec 1.5
	Iss string
	Hld string
	Bnd map[string]bound.Bound
	Prv string // "" for a root
	Exp int64
	Nnc string
}

// Call is a parsed call.
type Call struct {
	Raw   wire.Object
	ID    string // identity hash
	CID   string // the id member
	Chain []*Writ
	From  string
	Op    string
	Args  map[string]any
}

// Leaf returns the last writ of the chain.
func (c *Call) Leaf() *Writ { return c.Chain[len(c.Chain)-1] }

// Standing reports whether the call is a standing (sys/) call.
func (c *Call) Standing() bool { return strings.HasPrefix(c.Op, "sys/") }

// TallyErr is the err member of a tally.
type TallyErr struct {
	Code string
	Ref  string
}

// Tally is a parsed tally.
type Tally struct {
	Raw  wire.Object
	ID   string
	Call string
	Writ string
	Op   string
	Acc  int64
	St   string
	Err  *TallyErr
	Out  string // "" when null
	Used map[string]int64
	Rev  *int64 // until, nil when null
	Sub  []*Tally
	Wrt  []*Writ
}

// Revoke is a parsed revoke.
type Revoke struct {
	Raw   wire.Object
	ID    string
	Writ  string // hash or "*"
	Iss   string
	Chain []*Writ
}

// checkHeader runs spec 6.1 steps 1 to 4 on an object: size, canonical form,
// version, type, crit. Returns the canonical bytes.
func checkHeader(obj wire.Object, typ string, maxBytes int) ([]byte, error) {
	c, err := jcs.Marshal(obj)
	if err != nil {
		return nil, fail(Noncanonical, "%v", err)
	}
	if len(c) > maxBytes {
		return nil, fail(TooLarge, "%s is %d bytes, limit %d", typ, len(c), maxBytes)
	}
	v, ok := obj["v"].(json.Number)
	if !ok || v.String() != "1" {
		if vi, ok2 := obj["v"].(int); ok2 && vi == 1 {
			// tolerated for programmatically built objects
		} else {
			return nil, fail(UnsupportedVersion, "v is %v", obj["v"])
		}
	}
	if t, _ := obj["typ"].(string); t != typ {
		return nil, fail(WrongType, "typ is %q, want %q", t, typ)
	}
	if crit, present := obj["crit"]; present {
		arr, ok := crit.([]any)
		if !ok {
			return nil, fail(Malformed, "crit must be an array")
		}
		for _, e := range arr {
			name, ok := e.(string)
			if !ok {
				return nil, fail(Malformed, "crit entries must be strings")
			}
			if !known[typ][name] {
				return nil, fail(UnsupportedCritical, "crit names %q", name)
			}
			if _, ok := obj[name]; !ok {
				return nil, fail(Malformed, "crit names absent member %q", name)
			}
		}
	}
	return c, nil
}

// BoundReason maps a bound.Parse error to its reason code.
func BoundReason(err error) Reason {
	switch {
	case errors.Is(err, bound.ErrUnknownType):
		return UnknownBound
	case errors.Is(err, bound.ErrValue):
		return Noncanonical
	default:
		return Malformed
	}
}

// checkB64 enforces spec 1.1 rule 5 on a binary member: base64url alphabet,
// no padding, and an encoding that re-encodes to itself (noncanonical
// otherwise), then the expected decoded length (malformed otherwise). want 0
// means "at least MinRandom characters".
func checkB64(name, s string, want int) error {
	raw, err := wire.B64.DecodeString(s)
	if err != nil || !b64Re.MatchString(s) || wire.B64.EncodeToString(raw) != s {
		return fail(Noncanonical, "%s is not canonical base64url", name)
	}
	if want == 0 {
		if len(raw) < 16 {
			return fail(Malformed, "%s must encode at least 16 bytes", name)
		}
		return nil
	}
	if len(raw) != want {
		return fail(Malformed, "%s must encode %d bytes", name, want)
	}
	return nil
}

func getInt(obj wire.Object, name string) (int64, bool) {
	switch x := obj[name].(type) {
	case json.Number:
		n, err := x.Int64()
		return n, err == nil
	case int:
		return int64(x), true
	case int64:
		return x, true
	}
	return 0, false
}

func checkKey(did string) error {
	if _, err := keys.PublicKeyFromDID(did); err != nil {
		return fail(BadKey, "%v", err)
	}
	return nil
}

// ParseWrit validates structure (spec 6.1 steps 1 to 5) and signature (step 6).
func ParseWrit(obj wire.Object) (*Writ, error) {
	c, err := checkHeader(obj, "writ", MaxWritBytes)
	if err != nil {
		return nil, err
	}
	w := &Writ{Raw: obj, ID: wire.HashBytes(c)}
	var ok bool
	if w.Iss, ok = obj["iss"].(string); !ok {
		return nil, fail(Malformed, "iss must be a string")
	}
	if w.Hld, ok = obj["hld"].(string); !ok {
		return nil, fail(Malformed, "hld must be a string")
	}
	bnd, ok := obj["bnd"].(map[string]any)
	if !ok {
		return nil, fail(Malformed, "bnd must be an object")
	}
	w.Bnd = map[string]bound.Bound{}
	for name, v := range bnd {
		b, err := bound.Parse(v)
		if err != nil {
			return nil, fail(BoundReason(err), "bound %q: %v", name, err)
		}
		w.Bnd[name] = b
	}
	act, ok := w.Bnd["act"]
	if !ok || act.T != "prefix" {
		return nil, fail(Malformed, "bnd.act must be present with type prefix")
	}
	if h, ok := w.Bnd["hld"]; ok {
		if h.T != "set" {
			return nil, fail(Malformed, "bnd.hld must have type set")
		}
		for _, e := range h.Set {
			if e.IsInt {
				return nil, fail(Malformed, "bnd.hld elements must be keys")
			}
			if err := checkKey(e.Str); err != nil {
				return nil, err
			}
		}
	}
	if d, ok := w.Bnd["depth"]; ok && d.T != "max" {
		return nil, fail(Malformed, "bnd.depth must have type max")
	}
	prv, present := obj["prv"]
	if !present {
		return nil, fail(Malformed, "prv is required (null for a root)")
	}
	if prv != nil {
		s, ok := prv.(string)
		if !ok {
			return nil, fail(Malformed, "prv must be null or a hash")
		}
		if err := checkB64("prv", s, 32); err != nil {
			return nil, err
		}
		w.Prv = s
	}
	if w.Exp, ok = getInt(obj, "exp"); !ok {
		return nil, fail(Malformed, "exp must be an integer")
	}
	if w.Nnc, ok = obj["nnc"].(string); !ok {
		return nil, fail(Malformed, "nnc must be a string")
	}
	if err := checkB64("nnc", w.Nnc, 0); err != nil {
		return nil, err
	}
	if err := checkKey(w.Iss); err != nil {
		return nil, err
	}
	if err := checkKey(w.Hld); err != nil {
		return nil, err
	}
	if err := checkSig(obj, w.Iss); err != nil {
		return nil, err
	}
	return w, nil
}

func checkSig(obj wire.Object, signer string) error {
	s, ok := obj["sig"].(string)
	if !ok {
		return fail(Malformed, "sig must be a string")
	}
	if err := checkB64("sig", s, 64); err != nil {
		return err
	}
	if err := wire.VerifySig(obj, signer); err != nil {
		return fail(BadSignature, "%v", err)
	}
	return nil
}

// ParseChain parses every writ in an array. It does not check attenuation.
func ParseChain(v any) ([]*Writ, error) {
	arr, ok := v.([]any)
	if !ok {
		return nil, fail(Malformed, "chain must be an array")
	}
	if len(arr) > MaxChain {
		return nil, fail(TooLarge, "chain has %d writs, limit %d", len(arr), MaxChain)
	}
	var chain []*Writ
	for i, e := range arr {
		obj, ok := e.(map[string]any)
		if !ok {
			return nil, fail(Malformed, "chain[%d] is not an object", i)
		}
		w, err := ParseWrit(obj)
		if err != nil {
			return nil, err
		}
		chain = append(chain, w)
	}
	return chain, nil
}

// ParseCall validates a call's structure and signature and parses its chain
// (each writ's structure and signature, not attenuation).
func ParseCall(obj wire.Object) (*Call, error) {
	c, err := checkHeader(obj, "call", MaxCallBytes)
	if err != nil {
		return nil, err
	}
	k := &Call{Raw: obj, ID: wire.HashBytes(c)}
	var ok bool
	if k.CID, ok = obj["id"].(string); !ok {
		return nil, fail(Malformed, "id must be a string")
	}
	if err := checkB64("id", k.CID, 0); err != nil {
		return nil, err
	}
	if k.Chain, err = ParseChain(obj["chain"]); err != nil {
		return nil, err
	}
	if len(k.Chain) == 0 {
		return nil, fail(Malformed, "chain must not be empty")
	}
	if k.From, ok = obj["from"].(string); !ok {
		return nil, fail(Malformed, "from must be a string")
	}
	if k.Op, ok = obj["op"].(string); !ok {
		return nil, fail(Malformed, "op must be a string")
	}
	if k.Args, ok = obj["args"].(map[string]any); !ok {
		return nil, fail(Malformed, "args must be an object")
	}
	if err := checkKey(k.From); err != nil {
		return nil, err
	}
	if err := checkSig(obj, k.From); err != nil {
		return nil, err
	}
	return k, nil
}

// ParseTally validates a tally's structure and its signature under signer,
// which the caller derives from the writ the tally names. Sub-tallies and
// wrt entries are parsed structurally; sub-tally signatures are verified by
// VerifyTally once their writs are known.
func ParseTally(obj wire.Object, signer string) (*Tally, error) {
	c, err := checkHeader(obj, "tally", MaxTallyBytes)
	if err != nil {
		return nil, err
	}
	t := &Tally{Raw: obj, ID: wire.HashBytes(c)}
	var ok bool
	if t.Call, ok = obj["call"].(string); !ok {
		return nil, fail(Malformed, "call must be a hash")
	}
	if err := checkB64("call", t.Call, 32); err != nil {
		return nil, err
	}
	if t.Writ, ok = obj["writ"].(string); !ok {
		return nil, fail(Malformed, "writ must be a hash")
	}
	if err := checkB64("writ", t.Writ, 32); err != nil {
		return nil, err
	}
	if t.Op, ok = obj["op"].(string); !ok {
		return nil, fail(Malformed, "op must be a string")
	}
	if t.Acc, ok = getInt(obj, "acc"); !ok {
		return nil, fail(Malformed, "acc must be an integer")
	}
	if t.St, ok = obj["st"].(string); !ok || (t.St != "ok" && t.St != "failed" && t.St != "canceled" && t.St != "pending") {
		return nil, fail(Malformed, "st must be ok, failed, canceled, or pending")
	}
	errv, present := obj["err"]
	if !present {
		return nil, fail(Malformed, "err is required (null when ok)")
	}
	if errv != nil {
		m, ok := errv.(map[string]any)
		if !ok {
			return nil, fail(Malformed, "err must be null or an object")
		}
		code, ok := m["code"].(string)
		if !ok {
			return nil, fail(Malformed, "err.code must be a string")
		}
		t.Err = &TallyErr{Code: code}
		if ref, ok := m["ref"]; ok {
			s, ok := ref.(string)
			if !ok {
				return nil, fail(Malformed, "err.ref must be a hash")
			}
			if err := checkB64("err.ref", s, 32); err != nil {
				return nil, err
			}
			t.Err.Ref = s
		}
	}
	if t.St == "ok" && t.Err != nil {
		return nil, fail(Malformed, "err must be null when st is ok")
	}
	if t.St != "ok" && t.Err == nil {
		return nil, fail(Malformed, "err must be present when st is not ok")
	}
	out, present := obj["out"]
	if !present {
		return nil, fail(Malformed, "out is required (null when none)")
	}
	if out != nil {
		if t.Out, ok = out.(string); !ok {
			return nil, fail(Malformed, "out must be null or a hash")
		}
		if err := checkB64("out", t.Out, 32); err != nil {
			return nil, err
		}
	}
	used, ok := obj["used"].(map[string]any)
	if !ok {
		return nil, fail(Malformed, "used must be an object")
	}
	t.Used = map[string]int64{}
	for name, v := range used {
		n, ok := getInt(used, name)
		if !ok || n < 0 {
			return nil, fail(Malformed, "used.%s must be a non-negative integer (%v)", name, v)
		}
		t.Used[name] = n
	}
	rev, present := obj["rev"]
	if !present {
		return nil, fail(Malformed, "rev is required (null when not reversible)")
	}
	if rev != nil {
		m, ok := rev.(map[string]any)
		if !ok {
			return nil, fail(Malformed, "rev must be null or an object")
		}
		until, ok := getInt(m, "until")
		if !ok {
			return nil, fail(Malformed, "rev.until must be an integer")
		}
		t.Rev = &until
	}
	if t.St == "pending" && (len(t.Used) != 0 || t.Rev != nil || t.Out != "") {
		return nil, fail(Malformed, "a pending tally has empty used, null rev, null out")
	}
	wrt, ok := obj["wrt"].([]any)
	if !ok {
		return nil, fail(Malformed, "wrt must be an array")
	}
	for i, e := range wrt {
		o, ok := e.(map[string]any)
		if !ok {
			return nil, fail(Malformed, "wrt[%d] is not an object", i)
		}
		w, err := ParseWrit(o)
		if err != nil {
			return nil, err
		}
		t.Wrt = append(t.Wrt, w)
	}
	sub, ok := obj["sub"].([]any)
	if !ok {
		return nil, fail(Malformed, "sub must be an array")
	}
	for i, e := range sub {
		o, ok := e.(map[string]any)
		if !ok {
			return nil, fail(Malformed, "sub[%d] is not an object", i)
		}
		// Signer of a sub-tally is the hld of the writ it names, which must be in wrt.
		wh, _ := o["writ"].(string)
		var signerSub string
		for _, w := range t.Wrt {
			if w.ID == wh {
				signerSub = w.Hld
			}
		}
		if signerSub == "" {
			return nil, fail(SubUnmatched, "sub[%d] names writ %s absent from wrt", i, wh)
		}
		st, err := ParseTally(o, signerSub)
		if err != nil {
			return nil, err
		}
		t.Sub = append(t.Sub, st)
	}
	if err := checkSig(obj, signer); err != nil {
		return nil, err
	}
	return t, nil
}

// ParseRevoke validates a revoke's structure and signature.
func ParseRevoke(obj wire.Object) (*Revoke, error) {
	c, err := checkHeader(obj, "revoke", MaxRevokeBytes)
	if err != nil {
		return nil, err
	}
	r := &Revoke{Raw: obj, ID: wire.HashBytes(c)}
	var ok bool
	if r.Writ, ok = obj["writ"].(string); !ok {
		return nil, fail(Malformed, "writ must be a hash or \"*\"")
	}
	if r.Writ != "*" {
		if err := checkB64("writ", r.Writ, 32); err != nil {
			return nil, err
		}
	}
	if r.Iss, ok = obj["iss"].(string); !ok {
		return nil, fail(Malformed, "iss must be a string")
	}
	if r.Chain, err = ParseChain(obj["chain"]); err != nil {
		return nil, err
	}
	if err := checkKey(r.Iss); err != nil {
		return nil, err
	}
	if err := checkSig(obj, r.Iss); err != nil {
		return nil, err
	}
	return r, nil
}
